Add Custom-Rocm-Kernel-Skill (#3308)
huggingface/blog: rocm-kernel-skill.md

---
title: "ROCm Kernel Skill: Custom Triton Kernels for AMD, from Datacenter to Desktop"
thumbnail: /blog/assets/rocm-kernel-skill/meme.png
authors:
- user: Phlimos
- user: vivienfanghua
---

# ROCm Kernel Skill: Custom Triton Kernels for AMD, from Datacenter to Desktop

![ROCm Kernel Skill](assets/rocm-kernel-skill/meme.png)

> tl;dr: We built an agent skill that teaches coding agents how to write optimized Triton kernels for AMD GPUs. We tested it on two fundamentally different architectures — **MI355X** (CDNA3+, datacenter) and **R9700** (RDNA4, desktop) — with four kernel types. All passed correctness checks. R9700 reached **3.97x** peak kernel speedup with **79.5%** bandwidth utilization; MI355X delivered **25%** end-to-end speedup on LTX-Video. No CUDA required.

Reviewer: Nice opener, but this feels a bit overloaded for a `tl;dr`. I think this lands more cleanly as a short lead sentence plus bullets.

```suggestion
tl;dr: We built an agent skill that helps coding agents write optimized Triton kernels for AMD GPUs.

- Tested on **MI355X** (CDNA3+, datacenter) and **R9700** (RDNA4, desktop)
- Covers four kernel types, all passing correctness checks
- Up to **3.97x** peak kernel speedup on R9700
- **25%** end-to-end speedup on LTX-Video on MI355X
- No CUDA required
```

We recently released a [CUDA kernel skill](https://huggingface.co/blog/custom-cuda-kernels-agent-skills) that teaches coding agents to write production NVIDIA kernels. Now we are bringing the same capability to AMD — but this is not a port. AMD kernel development uses a different language, faces different constraints, and targets fundamentally different hardware. It required a skill built from scratch.

## Why AMD needs its own kernel skill

You might assume an AMD kernel skill is the CUDA skill with a few parameters changed. It is not. Three fundamental differences forced us to build a separate skill.

> **Different language.** The CUDA skill generates C/C++ kernel source with PyTorch C++ bindings. On AMD, we chose [Triton](https://triton-lang.org/) — a Python-native kernel language that compiles to AMD GCN/RDNA ISA via ROCm. No C++ bindings needed, no `torch_binding.cpp`, no `build.toml` with cuda-capabilities. The entire development workflow is Python end to end.

Reviewer: This is useful, but I’d trim a little jargon here so the contrast is easier to absorb on first read.

```suggestion
**Different language.** The CUDA skill generates C/C++ kernels with PyTorch bindings. The ROCm skill uses [Triton](https://triton-lang.org/), a Python-native kernel language that compiles through ROCm. That means a fully Python workflow, with no C++ bindings or `torch_binding.cpp` step.
```

**Different constraints.** ROCm's Triton has restrictions that do not exist on NVIDIA:

```python
# These work on NVIDIA Triton, but FAIL on ROCm
tl.libdevice.tanh(x)    # Not available
tl.math.tanh(x)          # Not available

# ROCm requires manual implementation
e2x = tl.exp(2.0 * x)
tanh_x = (e2x - 1.0) / (e2x + 1.0)
```

Autotuning `BLOCK_D` — perfectly safe on NVIDIA — produces **silently wrong results** on ROCm because the autotuner may pick a block size smaller than the row width, causing partial row processing. The skill encodes these pitfalls so the agent avoids them.

**Different hardware.** The CUDA skill targets H100, A100, and T4 — variations of the same NVIDIA architecture family. The ROCm skill targets two architectures that have almost nothing in common:

| | MI355X (CDNA3+) | R9700 (RDNA4) |
| :---- | :---: | :---: |
| Target | Datacenter | Desktop |
| Wavefront | Wave64 | Wave32 |
| XCDs (Chiplets) | 32 | 1 (monolithic) |
| LDS per CU | 160 KB | 64 KB |
| Memory BW | 8 TB/s | 608 GB/s |
| XCD Swizzle | Mandatory for GEMM | Not needed |

> Wave64 vs Wave32 changes how reductions work. XCD Swizzle is mandatory on MI355X — without it, GEMM blocks cluster on a few chiplets and waste 90%+ of the GPU — but is meaningless on R9700. One skill must teach the agent to handle both.

Reviewer: I’d simplify this a bit. The current version is accurate, but it asks the reader to hold a lot of detail in one sentence.

```suggestion
Wave64 vs Wave32 changes how reductions work. MI355X also needs XCD swizzle for GEMM to spread work across chiplets, while R9700 does not. The skill has to teach the agent both patterns.
```

### MI355X (CDNA3+, Datacenter)

**Hardware**: AMD Instinct MI355X, 288 GB HBM3e, ROCm 7.0

#### Isolated kernel micro-benchmarks

![MI355X kernel micro-benchmarks](assets/rocm-kernel-skill/MI355x_kernels.png)

RMSNorm bandwidth: 3,546 GB/s achieved (44% of MI355X theoretical 8 TB/s).

<details>
<summary>Table</summary>

| Kernel | Avg Speedup | Best Speedup | Best Config |
| :---- | :---: | :---: | :---- |
| RMSNorm | **1.73x** | 2.46x | [4×4096×3072] |
| RoPE 3D | **1.21x** | 1.54x | [2×4096×16×128] |
| GEGLU | **1.45x** | 2.20x | [4×4096×8192] |
| AdaLN | **2.28x** | 2.87x | [4×4096×3072] |

**Detailed RMSNorm results:**

| Shape | Triton (ms) | PyTorch (ms) | Speedup |
| :---- | :---: | :---: | :---: |
| [1×1024×2048] | 0.029 | 0.037 | **1.29x** |
| [2×1024×2048] | 0.029 | 0.044 | **1.48x** |
| [4×1024×2048] | 0.033 | 0.051 | **1.57x** |
| [1×4096×2048] | 0.032 | 0.050 | **1.56x** |
| [2×4096×3072] | 0.042 | 0.085 | **2.02x** |
| [1×8192×2048] | 0.040 | 0.067 | **1.69x** |
| [4×4096×3072] | 0.056 | 0.138 | **2.46x** |

</details>

#### End-to-end: LTX-Video (25 frames, 480×704, 30 steps)

> A 25% end-to-end speedup from custom Triton kernels alone. Combined with `torch.compile`, the pipeline reaches 1.52x over baseline. Peak memory stays at 18.58 GB in all configurations.

Reviewer: Slightly tighter phrasing would make this takeaway punchier.

```suggestion
Custom Triton kernels alone yield a 25% end-to-end speedup. With `torch.compile`, the pipeline reaches 1.52x over baseline, while peak memory stays flat at 18.58 GB.
```
